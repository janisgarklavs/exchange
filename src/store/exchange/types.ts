export interface Exchange {
  base: string;
  date: string;
  rates: Record<string, number>;
}

export type ApiResponse = Record<string, any>;

export enum ExchangeActionTypes {
  FETCH_REQUEST = "@@exchange/FETCH_REQUEST",
  FETCH_SUCCESS = "@@exchange/FETCH_SUCCESS",
  FETCH_ERROR = "@@exchange/FETCH_ERROR"
}

export interface ExchangeState {
  readonly loading: boolean;
  readonly data?: Exchange;
  readonly errors?: string;
}
