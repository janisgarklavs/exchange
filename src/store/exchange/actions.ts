import { action } from "typesafe-actions";
import { ExchangeActionTypes, Exchange } from "./types";
import { Dispatch } from "redux";

export const EXCHANGE_ENDPOINT = "https://api.exchangeratesapi.io/latest";

export const fetchRequestStart = () =>
  action(ExchangeActionTypes.FETCH_REQUEST);

export const fetchSuccess = (data: Exchange) =>
  action(ExchangeActionTypes.FETCH_SUCCESS, data);
export const fetchError = (msg: string) =>
  action(ExchangeActionTypes.FETCH_ERROR, msg);

export const fetchRequest = (): any => {
  return async (dispatch: Dispatch) => {
    dispatch(fetchRequestStart());

    try {
      const res = await fetch(EXCHANGE_ENDPOINT);
      const data = await res.json();
      if (!res.ok) {
        throw Error(data.error);
      }
      dispatch(fetchSuccess(data));
    } catch (err) {
      dispatch(fetchError(err.message));
    }
  };
};
