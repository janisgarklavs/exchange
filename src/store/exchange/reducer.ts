import { Reducer } from "redux";
import { ExchangeState, ExchangeActionTypes } from "./types";

const initialState: ExchangeState = {
  data: undefined,
  errors: undefined,
  loading: false
};

const reducer: Reducer<ExchangeState> = (state = initialState, action) => {
  switch (action.type) {
    case ExchangeActionTypes.FETCH_REQUEST: {
      return { ...state, loading: true };
    }
    case ExchangeActionTypes.FETCH_SUCCESS: {
      return {
        ...state,
        loading: false,
        data: action.payload,
        errors: ""
      };
    }
    case ExchangeActionTypes.FETCH_ERROR: {
      return {
        ...state,
        loading: false,
        errors: action.payload
      };
    }
    default: {
      return state;
    }
  }
};

export { reducer as exchangeReducer };
