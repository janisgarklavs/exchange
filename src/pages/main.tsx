import React from "react";
import { Dispatch } from "redux";
import { connect } from "react-redux";

import { ApplicationState } from "../store";
import { fetchRequest } from "../store/exchange/actions";
import { Exchange as IExchange } from "../store/exchange/types";

import { Exchange } from "../components/Exchange";
import { FetchButton } from "../components/FetchButton";
import { Loader } from "../components/Loader";

import { Card, CardHeader, CardBody } from "../components/Card";

interface PropsFromState {
  loading: boolean;
  data?: IExchange;
  errors?: string;
}

interface PropsFromDispatch {
  fetchRequest: typeof fetchRequest;
}

export class MainPage extends React.Component<
  PropsFromState & PropsFromDispatch
> {
  private handleClick = () => {
    this.props.fetchRequest();
  };

  componentDidUpdate(): void {
    const { errors } = this.props;
    if (errors) {
      alert(errors);
    }
  }

  render() {
    const { data, errors, loading } = this.props;

    return (
      <Card>
        <CardHeader>
          <div style={{ fontSize: 20, marginBottom: 12, textAlign: "center" }}>
            Latest exchange rates
          </div>
          <FetchButton loading={loading} clickHandler={this.handleClick} />

          {errors && (
            <span
              className="error"
              style={{ color: "red", marginTop: 12, textAlign: "center" }}
            >
              Something went wrong!
            </span>
          )}
        </CardHeader>
        <CardBody>
          {loading && <Loader />}
          {data && !errors && !loading && <Exchange data={data} />}
        </CardBody>
      </Card>
    );
  }
}

const mapStateToProps = ({ exchange }: ApplicationState) => ({
  loading: exchange.loading,
  errors: exchange.errors,
  data: exchange.data
});

const mapDispatchToProps = (dispatch: any) => ({
  fetchRequest: () => dispatch(fetchRequest())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MainPage);
