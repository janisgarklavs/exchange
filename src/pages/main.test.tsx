import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { MainPage } from "./main";
import { fetchRequest } from "../store/exchange/actions";

describe("<MainPage />", () => {
  test("it renders correctly", () => {
    const mockedState = {
      loading: false,
      data: undefined,
      errors: undefined
    };

    const component = shallow(
      <MainPage fetchRequest={fetchRequest} {...mockedState} />
    );

    expect(toJson(component)).toMatchSnapshot();
    expect(component.find("Loader").length).toBe(0);
    expect(component.find("Exchange").length).toBe(0);
    expect(component.find("span.error").length).toBe(0);
  });

  test("it shows error message in case of error prop", () => {
    const mockedState = {
      loading: false,
      data: undefined,
      errors: "Here is some error"
    };

    const component = shallow(
      <MainPage fetchRequest={fetchRequest} {...mockedState} />
    );
    expect(component.find("Loader").length).toBe(0);
    expect(component.find("Exchange").length).toBe(0);
    expect(component.find("span.error").text()).toEqual(
      "Something went wrong!"
    );
  });

  test("it shows loader while loading prop is truthy", () => {
    const mockedState = {
      loading: true,
      data: undefined,
      errors: "Here is some error"
    };

    const component = shallow(
      <MainPage fetchRequest={fetchRequest} {...mockedState} />
    );
    expect(component.find("Loader").length).toBe(1);
    expect(component.find("Exchange").length).toBe(0);
  });

  test("it shows Exchange component when received data", () => {
    const mockedState = {
      loading: false,
      data: { base: "EUR", date: "2019-01-01", rates: { USD: 1.23 } },
      errors: ""
    };

    const component = shallow(
      <MainPage fetchRequest={fetchRequest} {...mockedState} />
    );
    expect(component.find("Loader").length).toBe(0);
    expect(component.find("Exchange").length).toBe(1);
    expect(component.find("span.error").length).toBe(0);
  });
});
