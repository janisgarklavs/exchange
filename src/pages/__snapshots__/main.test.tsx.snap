// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainPage /> it renders correctly 1`] = `
<Card>
  <Header>
    <div
      style={
        Object {
          "fontSize": 20,
          "marginBottom": 12,
          "textAlign": "center",
        }
      }
    >
      Latest exchange rates
    </div>
    <FetchButton
      clickHandler={[Function]}
      loading={false}
    />
  </Header>
  <Body />
</Card>
`;
